
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://microbiomedata.github.io/nmdc-schema/DataObject/" rel="canonical"/>
<link href="../assets/images/favicon.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.6.14" name="generator"/>
<title>Class: DataObject - NMDC Schema Documentation</title>
<link href="../assets/stylesheets/main.342714a4.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
</head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#class-dataobject">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="NMDC Schema Documentation" class="md-header__button md-logo" data-md-component="logo" href=".." title="NMDC Schema Documentation">
<img alt="logo" src="../images/NMDC_logo.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            NMDC Schema Documentation
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Class: DataObject
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/nmdc-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="NMDC Schema Documentation" class="md-nav__button md-logo" data-md-component="logo" href=".." title="NMDC Schema Documentation">
<img alt="logo" src="../images/NMDC_logo.png"/>
</a>
    NMDC Schema Documentation
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/nmdc-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
<span class="md-ellipsis">
    Overview
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../about/">
<span class="md-ellipsis">
    Funding and Acknowledgements
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../collaboration/">
<span class="md-ellipsis">
    Contributing
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../Metadata_Documentation_Overview/">
<span class="md-ellipsis">
    Metadata Standards Documentation
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../maintaining-the-schema/">
<span class="md-ellipsis">
    Maintaining the NMDC Schema
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../schema_element_deprecation_guide/">
<span class="md-ellipsis">
    Schema element deprecation guide
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../schema-validation/">
<span class="md-ellipsis">
    NMDC Schema Validation
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../identifiers/">
<span class="md-ellipsis">
    Identifiers
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../typecode-to-class-map/">
<span class="md-ellipsis">
    Typecode to class map
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../migrations/">
<span class="md-ellipsis">
    Version Migration Guides
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../visualizations/">
<span class="md-ellipsis">
    Visualizations
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../validation/">
<span class="md-ellipsis">
    Data Validation
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../developer-docs/">
<span class="md-ellipsis">
    Developer Docs
    
  </span>
</a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#inheritance">
<span class="md-ellipsis">
      Inheritance
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#slots">
<span class="md-ellipsis">
      Slots
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usages">
<span class="md-ellipsis">
      Usages
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#identifier-and-mapping-information">
<span class="md-ellipsis">
      Identifier and Mapping Information
    </span>
</a>
<nav aria-label="Identifier and Mapping Information" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-source">
<span class="md-ellipsis">
      Schema Source
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mappings">
<span class="md-ellipsis">
      Mappings
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#linkml-source">
<span class="md-ellipsis">
      LinkML Source
    </span>
</a>
<nav aria-label="LinkML Source" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#direct">
<span class="md-ellipsis">
      Direct
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#induced">
<span class="md-ellipsis">
      Induced
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="class-dataobject">Class: DataObject</h1>
<p><em>An object that primarily consists of symbols that represent information.   Files, records, and omics data are examples of data objects.</em></p>
<p>URI: <a href="https://w3id.org/nmdc/DataObject">nmdc:DataObject</a></p>
<div class="mermaid">classDiagram
  class DataObject
  click DataObject href "../DataObject"
    InformationObject &lt;|-- DataObject
      click InformationObject href "../InformationObject"



  DataObject : alternative_identifiers

  DataObject : compression_type

  DataObject : data_category

      DataObject --&gt; "1" DataCategoryEnum : data_category
    click DataCategoryEnum href "../DataCategoryEnum"

  DataObject : data_object_type

      DataObject --&gt; "1" FileTypeEnum : data_object_type
    click FileTypeEnum href "../FileTypeEnum"

  DataObject : description

  DataObject : file_size_bytes

  DataObject : id

  DataObject : in_manifest

      DataObject --&gt; "*" Manifest : in_manifest
    click Manifest href "../Manifest"

  DataObject : insdc_experiment_identifiers

  DataObject : md5_checksum

  DataObject : name

  DataObject : superseded_by

      DataObject --&gt; "0..1" WorkflowExecution : superseded_by
    click WorkflowExecution href "../WorkflowExecution"

  DataObject : type

  DataObject : url

  DataObject : was_generated_by

      DataObject --&gt; "0..1" DataEmitterProcess : was_generated_by
    click DataEmitterProcess href "../DataEmitterProcess"



</div>
<h2 id="inheritance">Inheritance</h2>
<ul>
<li><a href="../NamedThing/">NamedThing</a><ul>
<li><a href="../InformationObject/">InformationObject</a><ul>
<li><strong>DataObject</strong></li>
</ul>
</li>
</ul>
</li>
</ul>
<h2 id="slots">Slots</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>Cardinality and Range</th>
<th>Description</th>
<th>Inheritance</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../compression_type/">compression_type</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>If provided, specifies the compression type</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../data_category/">data_category</a></td>
<td>1 <br/> <a href="../DataCategoryEnum/">DataCategoryEnum</a></td>
<td>The category of the file, such as instrument data from data generation or pro...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../data_object_type/">data_object_type</a></td>
<td>1 <br/> <a href="../FileTypeEnum/">FileTypeEnum</a></td>
<td>The type of file represented by the data object</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../file_size_bytes/">file_size_bytes</a></td>
<td>0..1 <br/> <a href="../Bytes/">Bytes</a></td>
<td>Size of the file in bytes</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../insdc_experiment_identifiers/">insdc_experiment_identifiers</a></td>
<td>* <br/> <a href="../ExternalIdentifier/">ExternalIdentifier</a></td>
<td></td>
<td>direct</td>
</tr>
<tr>
<td><a href="../md5_checksum/">md5_checksum</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>MD5 checksum of file (pre-compressed)</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../url/">url</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td></td>
<td>direct</td>
</tr>
<tr>
<td><a href="../was_generated_by/">was_generated_by</a></td>
<td>0..1 <br/> <a href="../DataEmitterProcess/">DataEmitterProcess</a></td>
<td></td>
<td>direct</td>
</tr>
<tr>
<td><a href="../in_manifest/">in_manifest</a></td>
<td>* <br/> <a href="../Manifest/">Manifest</a></td>
<td>one or more combinations of other DataObjects that can be analyzed together</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../superseded_by/">superseded_by</a></td>
<td>0..1 <br/> <a href="../WorkflowExecution/">WorkflowExecution</a></td>
<td>Links a DataObject or WorkflowExecution record to a newer WorkflowExecution t...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../id/">id</a></td>
<td>1 <br/> <a href="../Uriorcurie/">Uriorcurie</a></td>
<td>A unique identifier for a thing</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../name/">name</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>A human readable label for an entity</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../description/">description</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>a human-readable description of a thing</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../alternative_identifiers/">alternative_identifiers</a></td>
<td>* <br/> <a href="../Uriorcurie/">Uriorcurie</a></td>
<td>A list of alternative identifiers for the entity</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../type/">type</a></td>
<td>1 <br/> <a href="../Uriorcurie/">Uriorcurie</a></td>
<td>the class_uri of the class that has been instantiated</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
</tbody>
</table>
<h2 id="usages">Usages</h2>
<table>
<thead>
<tr>
<th>used by</th>
<th>used in</th>
<th>type</th>
<th>used</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../NucleotideSequencing/">NucleotideSequencing</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MassSpectrometry/">MassSpectrometry</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../CalibrationInformation/">CalibrationInformation</a></td>
<td><a href="../calibration_object/">calibration_object</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../Database/">Database</a></td>
<td><a href="../data_object_set/">data_object_set</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetagenomeSequencing/">MetagenomeSequencing</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../DataGeneration/">DataGeneration</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../AnnotatingWorkflow/">AnnotatingWorkflow</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../AnnotatingWorkflow/">AnnotatingWorkflow</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetagenomeAnnotation/">MetagenomeAnnotation</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetagenomeAnnotation/">MetagenomeAnnotation</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetagenomeAssembly/">MetagenomeAssembly</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetagenomeAssembly/">MetagenomeAssembly</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetatranscriptomeAssembly/">MetatranscriptomeAssembly</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetatranscriptomeAssembly/">MetatranscriptomeAssembly</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetatranscriptomeAnnotation/">MetatranscriptomeAnnotation</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetatranscriptomeAnnotation/">MetatranscriptomeAnnotation</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetatranscriptomeExpressionAnalysis/">MetatranscriptomeExpressionAnalysis</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetatranscriptomeExpressionAnalysis/">MetatranscriptomeExpressionAnalysis</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MagsAnalysis/">MagsAnalysis</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MagsAnalysis/">MagsAnalysis</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../ReadQcAnalysis/">ReadQcAnalysis</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../ReadQcAnalysis/">ReadQcAnalysis</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../ReadBasedTaxonomyAnalysis/">ReadBasedTaxonomyAnalysis</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../ReadBasedTaxonomyAnalysis/">ReadBasedTaxonomyAnalysis</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetabolomicsAnalysis/">MetabolomicsAnalysis</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetabolomicsAnalysis/">MetabolomicsAnalysis</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetaproteomicsAnalysis/">MetaproteomicsAnalysis</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../MetaproteomicsAnalysis/">MetaproteomicsAnalysis</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../NomAnalysis/">NomAnalysis</a></td>
<td><a href="../has_input/">has_input</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
<tr>
<td><a href="../NomAnalysis/">NomAnalysis</a></td>
<td><a href="../has_output/">has_output</a></td>
<td>range</td>
<td><a href="./">DataObject</a></td>
</tr>
</tbody>
</table>
<h2 id="identifier-and-mapping-information">Identifier and Mapping Information</h2>
<h3 id="schema-source">Schema Source</h3>
<ul>
<li>from schema: <a href="https://w3id.org/nmdc/nmdc">https://w3id.org/nmdc/nmdc</a></li>
</ul>
<h2 id="mappings">Mappings</h2>
<table>
<thead>
<tr>
<th>Mapping Type</th>
<th>Mapped Value</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2 id="linkml-source">LinkML Source</h2>
<!-- TODO: investigate https://stackoverflow.com/questions/37606292/how-to-create-tabbed-code-blocks-in-mkdocs-or-sphinx -->
<h3 id="direct">Direct</h3>
<details>
<pre><code class="language-yaml">name: DataObject
description: An object that primarily consists of symbols that represent information.   Files,
  records, and omics data are examples of data objects.
from_schema: https://w3id.org/nmdc/nmdc
is_a: InformationObject
slots:
- compression_type
- data_category
- data_object_type
- file_size_bytes
- insdc_experiment_identifiers
- md5_checksum
- url
- was_generated_by
- in_manifest
- superseded_by
slot_usage:
  name:
    name: name
    required: true
  description:
    name: description
    required: true
  id:
    name: id
    required: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:dobj-{id_shoulder}-{id_blade}$'
      interpolated: true
  was_generated_by:
    name: was_generated_by
    range: DataEmitterProcess
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfmtex|wfnom|wfrbt|wfrqc)-{id_shoulder}-{id_blade}{id_version}$|{id_nmdc_prefix}:(omprc|dgms|dgns)-{id_shoulder}-{id_blade}$'
      interpolated: true
  data_object_type:
    name: data_object_type
    required: true
  data_category:
    name: data_category
    required: true
class_uri: nmdc:DataObject

</code></pre>
</details>
<h3 id="induced">Induced</h3>
<details>
<pre><code class="language-yaml">name: DataObject
description: An object that primarily consists of symbols that represent information.   Files,
  records, and omics data are examples of data objects.
from_schema: https://w3id.org/nmdc/nmdc
is_a: InformationObject
slot_usage:
  name:
    name: name
    required: true
  description:
    name: description
    required: true
  id:
    name: id
    required: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:dobj-{id_shoulder}-{id_blade}$'
      interpolated: true
  was_generated_by:
    name: was_generated_by
    range: DataEmitterProcess
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfmtex|wfnom|wfrbt|wfrqc)-{id_shoulder}-{id_blade}{id_version}$|{id_nmdc_prefix}:(omprc|dgms|dgns)-{id_shoulder}-{id_blade}$'
      interpolated: true
  data_object_type:
    name: data_object_type
    required: true
  data_category:
    name: data_category
    required: true
attributes:
  compression_type:
    name: compression_type
    description: If provided, specifies the compression type
    todos:
    - consider setting the range to an enum
    examples:
    - value: gzip
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: compression_type
    owner: DataObject
    domain_of:
    - DataObject
    range: string
  data_category:
    name: data_category
    description: The category of the file, such as instrument data from data generation
      or processed data from a workflow execution.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: data_category
    owner: DataObject
    domain_of:
    - DataObject
    range: DataCategoryEnum
    required: true
  data_object_type:
    name: data_object_type
    description: The type of file represented by the data object.
    examples:
    - value: Direct Infusion FT-ICR MS Analysis Results
    - value: GC-MS Metabolomics Results
    from_schema: https://w3id.org/nmdc/nmdc
    structured_aliases:
      data_object_type:
        literal_form: data_object_type
        predicate: EXACT_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
    rank: 1000
    alias: data_object_type
    owner: DataObject
    domain_of:
    - DataObject
    range: FileTypeEnum
    required: true
  file_size_bytes:
    name: file_size_bytes
    description: Size of the file in bytes
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: file_size_bytes
    owner: DataObject
    domain_of:
    - DataObject
    range: bytes
  insdc_experiment_identifiers:
    name: insdc_experiment_identifiers
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: external_database_identifiers
    mixins:
    - insdc_identifiers
    alias: insdc_experiment_identifiers
    owner: DataObject
    domain_of:
    - NucleotideSequencing
    - DataObject
    range: external_identifier
    multivalued: true
    pattern: ^insdc.sra:(E|D|S)RX[0-9]{6,}$
  md5_checksum:
    name: md5_checksum
    description: MD5 checksum of file (pre-compressed)
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: md5_checksum
    owner: DataObject
    domain_of:
    - DataObject
    range: string
  url:
    name: url
    notes:
    - See issue 207 - this clashes with the mixs field
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: url
    owner: DataObject
    domain_of:
    - ImageValue
    - Protocol
    - DataObject
    range: string
  was_generated_by:
    name: was_generated_by
    from_schema: https://w3id.org/nmdc/nmdc
    mappings:
    - prov:wasGeneratedBy
    rank: 1000
    alias: was_generated_by
    owner: DataObject
    domain_of:
    - FunctionalAnnotationAggMember
    - FunctionalAnnotation
    - DataObject
    range: DataEmitterProcess
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(wfmag|wfmb|wfmgan|wfmgas|wfmsa|wfmp|wfmt|wfmtan|wfmtas|wfmtex|wfnom|wfrbt|wfrqc)-{id_shoulder}-{id_blade}{id_version}$|{id_nmdc_prefix}:(omprc|dgms|dgns)-{id_shoulder}-{id_blade}$'
      interpolated: true
  in_manifest:
    name: in_manifest
    description: one or more combinations of other DataObjects that can be analyzed
      together
    comments:
    - A DataObject can be part of multiple manifests, for example, a DataObject could
      be part of a manifest for a single run of an instrument and a manifest for technical
      replicates of a single sample.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: in_manifest
    owner: DataObject
    domain_of:
    - DataObject
    range: Manifest
    multivalued: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:manif-{id_shoulder}-{id_blade}$'
      interpolated: true
  superseded_by:
    name: superseded_by
    description: Links a DataObject or WorkflowExecution record to a newer WorkflowExecution
      that  supersedes it, marking this record as outdated. The linked WorkflowExecution
      or resultant DataObjects should be used in favor of this record.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: superseded_by
    owner: DataObject
    domain_of:
    - DataObject
    - WorkflowExecution
    range: WorkflowExecution
  id:
    name: id
    description: A unique identifier for a thing. Must be either a CURIE shorthand
      for a URI or a complete URI
    notes:
    - 'abstracted pattern: prefix:typecode-authshoulder-blade(.version)?(_seqsuffix)?'
    - a minimum length of 3 characters is suggested for typecodes, but 1 or 2 characters
      will be accepted
    - typecodes must correspond 1:1 to a class in the NMDC schema. this will be checked
      via per-class id slot usage assertions
    - minting authority shoulders should probably be enumerated and checked in the
      pattern
    examples:
    - value: nmdc:mgmag-00-x012.1_7_c1
      description: https://github.com/microbiomedata/nmdc-schema/pull/499#discussion_r1018499248
    from_schema: https://w3id.org/nmdc/nmdc
    structured_aliases:
      workflow_execution_id:
        literal_form: workflow_execution_id
        predicate: NARROW_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
      data_object_id:
        literal_form: data_object_id
        predicate: NARROW_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
    rank: 1000
    identifier: true
    alias: id
    owner: DataObject
    domain_of:
    - NamedThing
    range: uriorcurie
    required: true
    pattern: ^[a-zA-Z0-9][a-zA-Z0-9_\.]+:[a-zA-Z0-9_][a-zA-Z0-9_\-\/\.,]*$
    structured_pattern:
      syntax: '{id_nmdc_prefix}:dobj-{id_shoulder}-{id_blade}$'
      interpolated: true
  name:
    name: name
    description: A human readable label for an entity
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: name
    owner: DataObject
    domain_of:
    - PersonValue
    - NamedThing
    - Protocol
    range: string
    required: true
  description:
    name: description
    description: a human-readable description of a thing
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    slot_uri: dcterms:description
    alias: description
    owner: DataObject
    domain_of:
    - ImageValue
    - NamedThing
    - Protocol
    range: string
    required: true
  alternative_identifiers:
    name: alternative_identifiers
    description: A list of alternative identifiers for the entity.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: alternative_identifiers
    owner: DataObject
    domain_of:
    - NamedThing
    - MetaboliteIdentification
    range: uriorcurie
    multivalued: true
    pattern: ^[a-zA-Z0-9][a-zA-Z0-9_\.]+:[a-zA-Z0-9_][a-zA-Z0-9_\-\/\.,\(\)\=\#]*$
  type:
    name: type
    description: the class_uri of the class that has been instantiated
    notes:
    - makes it easier to read example data files
    - required for polymorphic MongoDB collections
    examples:
    - value: nmdc:Biosample
    - value: nmdc:Study
    from_schema: https://w3id.org/nmdc/nmdc
    see_also:
    - https://github.com/microbiomedata/nmdc-schema/issues/1048
    - https://github.com/microbiomedata/nmdc-schema/issues/1233
    - https://github.com/microbiomedata/nmdc-schema/issues/248
    structured_aliases:
      workflow_execution_class:
        literal_form: workflow_execution_class
        predicate: NARROW_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
    rank: 1000
    slot_uri: rdf:type
    designates_type: true
    alias: type
    owner: DataObject
    domain_of:
    - EukEval
    - FunctionalAnnotationAggMember
    - PeptideQuantification
    - ProteinQuantification
    - GenomeFeature
    - FunctionalAnnotation
    - AttributeValue
    - NamedThing
    - OntologyRelation
    - FailureCategorization
    - Protocol
    - CreditAssociation
    - Doi
    - ProvenanceMetadata
    - MobilePhaseSegment
    - PortionOfSubstance
    - MagBin
    - MetaboliteIdentification
    range: uriorcurie
    required: true
class_uri: nmdc:DataObject

</code></pre>
</details>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link", "hide_page_title"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
<script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
<script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
<script src="../javascripts/tablesort.js"></script>
<script src="https://unpkg.com/mermaid@10.9.0/dist/mermaid.min.js"></script><script>mermaid.initialize({});</script></body>
</html>