
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://microbiomedata.github.io/nmdc-schema/MetatranscriptomeAssembly/" rel="canonical"/>
<link href="../assets/images/favicon.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.6.14" name="generator"/>
<title>Class: MetatranscriptomeAssembly - NMDC Schema Documentation</title>
<link href="../assets/stylesheets/main.342714a4.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
</head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#class-metatranscriptomeassembly">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="NMDC Schema Documentation" class="md-header__button md-logo" data-md-component="logo" href=".." title="NMDC Schema Documentation">
<img alt="logo" src="../images/NMDC_logo.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            NMDC Schema Documentation
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Class: MetatranscriptomeAssembly
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/nmdc-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="NMDC Schema Documentation" class="md-nav__button md-logo" data-md-component="logo" href=".." title="NMDC Schema Documentation">
<img alt="logo" src="../images/NMDC_logo.png"/>
</a>
    NMDC Schema Documentation
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/nmdc-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
<span class="md-ellipsis">
    Overview
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../about/">
<span class="md-ellipsis">
    Funding and Acknowledgements
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../collaboration/">
<span class="md-ellipsis">
    Contributing
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../Metadata_Documentation_Overview/">
<span class="md-ellipsis">
    Metadata Standards Documentation
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../maintaining-the-schema/">
<span class="md-ellipsis">
    Maintaining the NMDC Schema
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../schema_element_deprecation_guide/">
<span class="md-ellipsis">
    Schema element deprecation guide
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../schema-validation/">
<span class="md-ellipsis">
    NMDC Schema Validation
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../identifiers/">
<span class="md-ellipsis">
    Identifiers
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../typecode-to-class-map/">
<span class="md-ellipsis">
    Typecode to class map
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../migrations/">
<span class="md-ellipsis">
    Version Migration Guides
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../visualizations/">
<span class="md-ellipsis">
    Visualizations
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../validation/">
<span class="md-ellipsis">
    Data Validation
    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../developer-docs/">
<span class="md-ellipsis">
    Developer Docs
    
  </span>
</a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#inheritance">
<span class="md-ellipsis">
      Inheritance
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#slots">
<span class="md-ellipsis">
      Slots
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#identifier-and-mapping-information">
<span class="md-ellipsis">
      Identifier and Mapping Information
    </span>
</a>
<nav aria-label="Identifier and Mapping Information" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-source">
<span class="md-ellipsis">
      Schema Source
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mappings">
<span class="md-ellipsis">
      Mappings
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#linkml-source">
<span class="md-ellipsis">
      LinkML Source
    </span>
</a>
<nav aria-label="LinkML Source" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#direct">
<span class="md-ellipsis">
      Direct
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#induced">
<span class="md-ellipsis">
      Induced
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="class-metatranscriptomeassembly">Class: MetatranscriptomeAssembly</h1>
<p>URI: <a href="https://w3id.org/nmdc/MetatranscriptomeAssembly">nmdc:MetatranscriptomeAssembly</a></p>
<div class="mermaid">classDiagram
  class MetatranscriptomeAssembly
  click MetatranscriptomeAssembly href "../MetatranscriptomeAssembly"
    WorkflowExecution &lt;|-- MetatranscriptomeAssembly
      click WorkflowExecution href "../WorkflowExecution"



  MetatranscriptomeAssembly : alternative_identifiers

  MetatranscriptomeAssembly : asm_score

  MetatranscriptomeAssembly : contig_bp

  MetatranscriptomeAssembly : contigs

  MetatranscriptomeAssembly : ctg_l50

  MetatranscriptomeAssembly : ctg_l90

  MetatranscriptomeAssembly : ctg_logsum

  MetatranscriptomeAssembly : ctg_max

  MetatranscriptomeAssembly : ctg_n50

  MetatranscriptomeAssembly : ctg_n90

  MetatranscriptomeAssembly : ctg_powsum

  MetatranscriptomeAssembly : description

  MetatranscriptomeAssembly : end_date

  MetatranscriptomeAssembly : ended_at_time

  MetatranscriptomeAssembly : execution_resource

      MetatranscriptomeAssembly --&gt; "0..1" ExecutionResourceEnum : execution_resource
    click ExecutionResourceEnum href "../ExecutionResourceEnum"

  MetatranscriptomeAssembly : gap_pct

  MetatranscriptomeAssembly : gc_avg

  MetatranscriptomeAssembly : gc_std

  MetatranscriptomeAssembly : git_url

  MetatranscriptomeAssembly : has_failure_categorization

      MetatranscriptomeAssembly --&gt; "*" FailureCategorization : has_failure_categorization
    click FailureCategorization href "../FailureCategorization"

  MetatranscriptomeAssembly : has_input

      MetatranscriptomeAssembly --&gt; "1..*" DataObject : has_input
    click DataObject href "../DataObject"

  MetatranscriptomeAssembly : has_output

      MetatranscriptomeAssembly --&gt; "*" DataObject : has_output
    click DataObject href "../DataObject"

  MetatranscriptomeAssembly : id

  MetatranscriptomeAssembly : insdc_assembly_identifiers

  MetatranscriptomeAssembly : name

  MetatranscriptomeAssembly : num_aligned_reads

  MetatranscriptomeAssembly : num_input_reads

  MetatranscriptomeAssembly : processing_institution

      MetatranscriptomeAssembly --&gt; "1" ProcessingInstitutionEnum : processing_institution
    click ProcessingInstitutionEnum href "../ProcessingInstitutionEnum"

  MetatranscriptomeAssembly : processing_institution_workflow_metadata

  MetatranscriptomeAssembly : protocol_link

      MetatranscriptomeAssembly --&gt; "0..1" Protocol : protocol_link
    click Protocol href "../Protocol"

  MetatranscriptomeAssembly : qc_comment

  MetatranscriptomeAssembly : qc_status

      MetatranscriptomeAssembly --&gt; "0..1" StatusEnum : qc_status
    click StatusEnum href "../StatusEnum"

  MetatranscriptomeAssembly : scaf_bp

  MetatranscriptomeAssembly : scaf_l50

  MetatranscriptomeAssembly : scaf_l90

  MetatranscriptomeAssembly : scaf_l_gt50k

  MetatranscriptomeAssembly : scaf_logsum

  MetatranscriptomeAssembly : scaf_max

  MetatranscriptomeAssembly : scaf_n50

  MetatranscriptomeAssembly : scaf_n90

  MetatranscriptomeAssembly : scaf_n_gt50k

  MetatranscriptomeAssembly : scaf_pct_gt50k

  MetatranscriptomeAssembly : scaf_powsum

  MetatranscriptomeAssembly : scaffolds

  MetatranscriptomeAssembly : start_date

  MetatranscriptomeAssembly : started_at_time

  MetatranscriptomeAssembly : type

  MetatranscriptomeAssembly : version

  MetatranscriptomeAssembly : was_informed_by

      MetatranscriptomeAssembly --&gt; "1..*" NucleotideSequencing : was_informed_by
    click NucleotideSequencing href "../NucleotideSequencing"



</div>
<h2 id="inheritance">Inheritance</h2>
<ul>
<li><a href="../NamedThing/">NamedThing</a><ul>
<li><a href="../PlannedProcess/">PlannedProcess</a><ul>
<li><a href="../DataEmitterProcess/">DataEmitterProcess</a><ul>
<li><a href="../WorkflowExecution/">WorkflowExecution</a><ul>
<li><strong>MetatranscriptomeAssembly</strong></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2 id="slots">Slots</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>Cardinality and Range</th>
<th>Description</th>
<th>Inheritance</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../asm_score/">asm_score</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>A score for comparing metagenomic assembly quality from same sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaffolds/">scaffolds</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Total sequence count of all scaffolds</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_logsum/">scaf_logsum</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>The sum of the (length*log(length)) of all scaffolds, times some constant</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_powsum/">scaf_powsum</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Powersum of all scaffolds is the same as logsum except that it uses the sum o...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_max/">scaf_max</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Maximum scaffold length</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_bp/">scaf_bp</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Total size in bp of all scaffolds</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_n50/">scaf_n50</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Given a set of scaffolds, each with its own length, the N50 count is defined ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_n90/">scaf_n90</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Given a set of scaffolds, each with its own length, the N90 count is defined ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_l50/">scaf_l50</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Given a set of scaffolds, the L50 is defined as the sequence length of the sh...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_l90/">scaf_l90</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>The L90 statistic is less than or equal to the L50 statistic; it is the lengt...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_n_gt50k/">scaf_n_gt50k</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Total sequence count of scaffolds greater than 50 KB</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_l_gt50k/">scaf_l_gt50k</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Total size in bp of all scaffolds greater than 50 KB</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../scaf_pct_gt50k/">scaf_pct_gt50k</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Total sequence size percentage of scaffolds greater than 50 KB</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../contigs/">contigs</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>The sum of the (length*log(length)) of all contigs, times some constant</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../contig_bp/">contig_bp</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Total size in bp of all contigs</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ctg_n50/">ctg_n50</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Given a set of contigs, each with its own length, the N50 count is defined as...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ctg_l50/">ctg_l50</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Given a set of contigs, the L50 is defined as the sequence length of the shor...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ctg_n90/">ctg_n90</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Given a set of contigs, each with its own length, the N90 count is defined as...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ctg_l90/">ctg_l90</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>The L90 statistic is less than or equal to the L50 statistic; it is the lengt...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ctg_logsum/">ctg_logsum</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Maximum contig length</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ctg_powsum/">ctg_powsum</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Powersum of all contigs is the same as logsum except that it uses the sum of ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ctg_max/">ctg_max</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Maximum contig length</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../gap_pct/">gap_pct</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>The gap size percentage of all scaffolds</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../gc_std/">gc_std</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Standard deviation of GC content of all contigs</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../gc_avg/">gc_avg</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Average of GC content of all contigs</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../num_input_reads/">num_input_reads</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>The sequence count number of input reads for assembly</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../num_aligned_reads/">num_aligned_reads</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>The sequence count number of input reads aligned to assembled contigs</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../insdc_assembly_identifiers/">insdc_assembly_identifiers</a></td>
<td>* <br/> <a href="../ExternalIdentifier/">ExternalIdentifier</a></td>
<td></td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ended_at_time/">ended_at_time</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td></td>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
</tr>
<tr>
<td><a href="../execution_resource/">execution_resource</a></td>
<td>0..1 <br/> <a href="../ExecutionResourceEnum/">ExecutionResourceEnum</a></td>
<td>The computing resource or facility where the workflow was executed</td>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
</tr>
<tr>
<td><a href="../git_url/">git_url</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>The url that points to the exact github location of a workflow</td>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
</tr>
<tr>
<td><a href="../started_at_time/">started_at_time</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td></td>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
</tr>
<tr>
<td><a href="../version/">version</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The NMDC release tag for a given workflow release used for data processing</td>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
</tr>
<tr>
<td><a href="../was_informed_by/">was_informed_by</a></td>
<td>1..* <br/> <a href="../NucleotideSequencing/">NucleotideSequencing</a></td>
<td>The primary DataGeneration subclass that the WorkflowExecution subclass depen...</td>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
</tr>
<tr>
<td><a href="../processing_institution_workflow_metadata/">processing_institution_workflow_metadata</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about how workflow results were generated when the processing is ...</td>
<td><a href="../WorkflowExecution/">WorkflowExecution</a></td>
</tr>
<tr>
<td><a href="../has_input/">has_input</a></td>
<td>1..* <br/> <a href="../DataObject/">DataObject</a></td>
<td>An input to a process</td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../has_output/">has_output</a></td>
<td>* <br/> <a href="../DataObject/">DataObject</a></td>
<td>An output from a process</td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../processing_institution/">processing_institution</a></td>
<td>1 <br/> <a href="../ProcessingInstitutionEnum/">ProcessingInstitutionEnum</a></td>
<td>The organization that processed the sample</td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../protocol_link/">protocol_link</a></td>
<td>0..1 <br/> <a href="../Protocol/">Protocol</a></td>
<td></td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../start_date/">start_date</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The date on which any process or activity was started</td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../end_date/">end_date</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The date on which any process or activity was ended</td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../qc_status/">qc_status</a></td>
<td>0..1 <br/> <a href="../StatusEnum/">StatusEnum</a></td>
<td>Stores information about the result of a process (ie the process of sequencin...</td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../qc_comment/">qc_comment</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Slot to store additional comments about laboratory or workflow output</td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../has_failure_categorization/">has_failure_categorization</a></td>
<td>* <br/> <a href="../FailureCategorization/">FailureCategorization</a></td>
<td></td>
<td><a href="../PlannedProcess/">PlannedProcess</a></td>
</tr>
<tr>
<td><a href="../id/">id</a></td>
<td>1 <br/> <a href="../Uriorcurie/">Uriorcurie</a></td>
<td>A unique identifier for a thing</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../name/">name</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A human readable label for an entity</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../description/">description</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>a human-readable description of a thing</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../alternative_identifiers/">alternative_identifiers</a></td>
<td>* <br/> <a href="../Uriorcurie/">Uriorcurie</a></td>
<td>A list of alternative identifiers for the entity</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
<tr>
<td><a href="../type/">type</a></td>
<td>1 <br/> <a href="../Uriorcurie/">Uriorcurie</a></td>
<td>the class_uri of the class that has been instantiated</td>
<td><a href="../NamedThing/">NamedThing</a></td>
</tr>
</tbody>
</table>
<h2 id="identifier-and-mapping-information">Identifier and Mapping Information</h2>
<h3 id="schema-source">Schema Source</h3>
<ul>
<li>from schema: <a href="https://w3id.org/nmdc/nmdc">https://w3id.org/nmdc/nmdc</a></li>
</ul>
<h2 id="mappings">Mappings</h2>
<table>
<thead>
<tr>
<th>Mapping Type</th>
<th>Mapped Value</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2 id="linkml-source">LinkML Source</h2>
<!-- TODO: investigate https://stackoverflow.com/questions/37606292/how-to-create-tabbed-code-blocks-in-mkdocs-or-sphinx -->
<h3 id="direct">Direct</h3>
<details>
<pre><code class="language-yaml">name: MetatranscriptomeAssembly
from_schema: https://w3id.org/nmdc/nmdc
is_a: WorkflowExecution
slots:
- asm_score
- scaffolds
- scaf_logsum
- scaf_powsum
- scaf_max
- scaf_bp
- scaf_n50
- scaf_n90
- scaf_l50
- scaf_l90
- scaf_n_gt50k
- scaf_l_gt50k
- scaf_pct_gt50k
- contigs
- contig_bp
- ctg_n50
- ctg_l50
- ctg_n90
- ctg_l90
- ctg_logsum
- ctg_powsum
- ctg_max
- gap_pct
- gc_std
- gc_avg
- num_input_reads
- num_aligned_reads
- insdc_assembly_identifiers
slot_usage:
  id:
    name: id
    required: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:wfmtas-{id_shoulder}-{id_blade}{id_version}$'
      interpolated: true
  was_informed_by:
    name: was_informed_by
    range: NucleotideSequencing
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(omprc|dgns)-{id_shoulder}-{id_blade}$'
      interpolated: true
class_uri: nmdc:MetatranscriptomeAssembly

</code></pre>
</details>
<h3 id="induced">Induced</h3>
<details>
<pre><code class="language-yaml">name: MetatranscriptomeAssembly
from_schema: https://w3id.org/nmdc/nmdc
is_a: WorkflowExecution
slot_usage:
  id:
    name: id
    required: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:wfmtas-{id_shoulder}-{id_blade}{id_version}$'
      interpolated: true
  was_informed_by:
    name: was_informed_by
    range: NucleotideSequencing
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(omprc|dgns)-{id_shoulder}-{id_blade}$'
      interpolated: true
attributes:
  asm_score:
    name: asm_score
    description: A score for comparing metagenomic assembly quality from same sample.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: asm_score
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaffolds:
    name: scaffolds
    description: Total sequence count of all scaffolds.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaffolds
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_logsum:
    name: scaf_logsum
    description: The sum of the (length*log(length)) of all scaffolds, times some
      constant.  Increase the contiguity, the score will increase
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_logsum
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_powsum:
    name: scaf_powsum
    description: Powersum of all scaffolds is the same as logsum except that it uses
      the sum of (length*(length^P)) for some power P (default P=0.25).
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_powsum
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_max:
    name: scaf_max
    description: Maximum scaffold length.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_max
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_bp:
    name: scaf_bp
    description: Total size in bp of all scaffolds.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_bp
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_n50:
    name: scaf_n50
    description: Given a set of scaffolds, each with its own length, the N50 count
      is defined as the smallest number of scaffolds whose length sum makes up half
      of genome size.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_n50
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_n90:
    name: scaf_n90
    description: Given a set of scaffolds, each with its own length, the N90 count
      is defined as the smallest number of scaffolds whose length sum makes up 90%
      of genome size.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_n90
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_l50:
    name: scaf_l50
    description: Given a set of scaffolds, the L50 is defined as the sequence length
      of the shortest scaffold at 50% of the total genome length.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_l50
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_l90:
    name: scaf_l90
    description: The L90 statistic is less than or equal to the L50 statistic; it
      is the length for which the collection of all scaffolds of that length or longer
      contains at least 90% of the sum of the lengths of all scaffolds.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_l90
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_n_gt50k:
    name: scaf_n_gt50k
    description: Total sequence count of scaffolds greater than 50 KB.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_n_gt50k
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_l_gt50k:
    name: scaf_l_gt50k
    description: Total size in bp of all scaffolds greater than 50 KB.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_l_gt50k
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  scaf_pct_gt50k:
    name: scaf_pct_gt50k
    description: Total sequence size percentage of scaffolds greater than 50 KB.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: scaf_pct_gt50k
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  contigs:
    name: contigs
    description: The sum of the (length*log(length)) of all contigs, times some constant.  Increase
      the contiguity, the score will increase
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: contigs
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  contig_bp:
    name: contig_bp
    description: Total size in bp of all contigs.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: contig_bp
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  ctg_n50:
    name: ctg_n50
    description: Given a set of contigs, each with its own length, the N50 count is
      defined as the smallest number_of_contigs whose length sum makes up half of
      genome size.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: ctg_n50
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  ctg_l50:
    name: ctg_l50
    description: Given a set of contigs, the L50 is defined as the sequence length
      of the shortest contig at 50% of the total genome length.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: ctg_l50
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  ctg_n90:
    name: ctg_n90
    description: Given a set of contigs, each with its own length, the N90 count is
      defined as the smallest number of contigs whose length sum makes up 90% of genome
      size.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: ctg_n90
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  ctg_l90:
    name: ctg_l90
    description: The L90 statistic is less than or equal to the L50 statistic; it
      is the length for which the collection of all contigs of that length or longer
      contains at least 90% of the sum of the lengths of all contigs.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: ctg_l90
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  ctg_logsum:
    name: ctg_logsum
    description: Maximum contig length.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: ctg_logsum
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  ctg_powsum:
    name: ctg_powsum
    description: Powersum of all contigs is the same as logsum except that it uses
      the sum of (length*(length^P)) for some power P (default P=0.25).
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: ctg_powsum
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  ctg_max:
    name: ctg_max
    description: Maximum contig length.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: ctg_max
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  gap_pct:
    name: gap_pct
    description: The gap size percentage of all scaffolds.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: gap_pct
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  gc_std:
    name: gc_std
    description: Standard deviation of GC content of all contigs.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: gc_std
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  gc_avg:
    name: gc_avg
    description: Average of GC content of all contigs.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: gc_avg
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  num_input_reads:
    name: num_input_reads
    description: The sequence count number of input reads for assembly.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: num_input_reads
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  num_aligned_reads:
    name: num_aligned_reads
    description: The sequence count number of input reads aligned to assembled contigs.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: metagenome_assembly_parameter
    alias: num_aligned_reads
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: float
  insdc_assembly_identifiers:
    name: insdc_assembly_identifiers
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    is_a: assembly_identifiers
    mixins:
    - insdc_identifiers
    alias: insdc_assembly_identifiers
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetagenomeAssembly
    - MetatranscriptomeAssembly
    range: external_identifier
    multivalued: true
    pattern: ^insdc.sra:[A-Z]+[0-9]+(\.[0-9]+)?$
  ended_at_time:
    name: ended_at_time
    notes:
    - 'The regex for ISO-8601 format was taken from here: https://www.myintervals.com/blog/2009/05/20/iso-8601-date-validation-that-doesnt-suck/
      It may not be complete, but it is good enough for now.'
    from_schema: https://w3id.org/nmdc/nmdc
    mappings:
    - prov:endedAtTime
    rank: 1000
    alias: ended_at_time
    owner: MetatranscriptomeAssembly
    domain_of:
    - WorkflowExecution
    range: string
    pattern: ^([\+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24\:?00)([\.,]\d+(?!:))?)?(\17[0-5]\d([\.,]\d+)?)?([zZ]|([\+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$
  execution_resource:
    name: execution_resource
    description: The computing resource or facility where the workflow was executed.
    examples:
    - value: NERSC-Cori
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: execution_resource
    owner: MetatranscriptomeAssembly
    domain_of:
    - WorkflowExecution
    range: ExecutionResourceEnum
  git_url:
    name: git_url
    description: The url that points to the exact github location of a workflow.
    examples:
    - value: https://github.com/microbiomedata/mg_annotation/releases/tag/0.1
    - value: https://github.com/microbiomedata/metaMS/blob/master/metaMS/gcmsWorkflow.py
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: git_url
    owner: MetatranscriptomeAssembly
    domain_of:
    - WorkflowExecution
    range: string
    required: true
  started_at_time:
    name: started_at_time
    notes:
    - 'The regex for ISO-8601 format was taken from here: https://www.myintervals.com/blog/2009/05/20/iso-8601-date-validation-that-doesnt-suck/
      It may not be complete, but it is good enough for now.'
    from_schema: https://w3id.org/nmdc/nmdc
    mappings:
    - prov:startedAtTime
    rank: 1000
    alias: started_at_time
    owner: MetatranscriptomeAssembly
    domain_of:
    - WorkflowExecution
    range: string
    required: true
    pattern: ^([\+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24\:?00)([\.,]\d+(?!:))?)?(\17[0-5]\d([\.,]\d+)?)?([zZ]|([\+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$
  version:
    name: version
    description: The NMDC release tag for a given workflow release used for data processing.
      If workflows are processed externally, as denoted by processing_institution,
      this value represents the best mapping between a processing institution's (e.g.,
      JGI) workflow metadata and a NMDC tagged release.
    examples:
    - value: v1.2.0
    from_schema: https://w3id.org/nmdc/nmdc
    broad_mappings:
    - NCIT:C182117
    rank: 1000
    alias: version
    owner: MetatranscriptomeAssembly
    domain_of:
    - WorkflowExecution
    range: string
  was_informed_by:
    name: was_informed_by
    description: The primary DataGeneration subclass that the WorkflowExecution subclass
      depends on.
    comments:
    - For version 1 of the proteomics workflow there are input files both from the
      NucleotideSequencing and MassSpectrometry, the MassSpectrometry record is considered
      the primary class to reference.
    from_schema: https://w3id.org/nmdc/nmdc
    structured_aliases:
      was_informed_by:
        literal_form: was_informed_by
        predicate: EXACT_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
    narrow_mappings:
    - prov:wasInformedBy
    rank: 1000
    alias: was_informed_by
    owner: MetatranscriptomeAssembly
    domain_of:
    - WorkflowExecution
    range: NucleotideSequencing
    required: true
    multivalued: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(omprc|dgns)-{id_shoulder}-{id_blade}$'
      interpolated: true
  processing_institution_workflow_metadata:
    name: processing_institution_workflow_metadata
    description: Information about how workflow results were generated when the processing
      is done by an external organziation (e.g., JGI) such as software tool name and
      version or pipeline name and version.
    examples:
    - value: metaspades v. 3.15.2
    - value: IMG Annotation Pipeline v.5.0.25
    from_schema: https://w3id.org/nmdc/nmdc
    mappings:
    - NCIT:C165211
    rank: 1000
    alias: processing_institution_workflow_metadata
    owner: MetatranscriptomeAssembly
    domain_of:
    - WorkflowExecution
    range: string
  has_input:
    name: has_input
    description: An input to a process.
    from_schema: https://w3id.org/nmdc/nmdc
    aliases:
    - input
    rank: 1000
    alias: has_input
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: DataObject
    required: true
    multivalued: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(dobj)-{id_shoulder}-{id_blade}$'
      interpolated: true
  has_output:
    name: has_output
    description: An output from a process.
    from_schema: https://w3id.org/nmdc/nmdc
    aliases:
    - output
    rank: 1000
    alias: has_output
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: DataObject
    multivalued: true
    structured_pattern:
      syntax: '{id_nmdc_prefix}:(dobj)-{id_shoulder}-{id_blade}$'
      interpolated: true
  processing_institution:
    name: processing_institution
    description: The organization that processed the sample.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: processing_institution
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: ProcessingInstitutionEnum
    required: true
  protocol_link:
    name: protocol_link
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: protocol_link
    owner: MetatranscriptomeAssembly
    domain_of:
    - Configuration
    - PlannedProcess
    - Study
    range: Protocol
  start_date:
    name: start_date
    description: The date on which any process or activity was started
    todos:
    - add date string validation pattern
    comments:
    - We are using string representations of dates until all components of our ecosystem
      can handle ISO 8610 dates
    - The date should be formatted as YYYY-MM-DD
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: start_date
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: string
  end_date:
    name: end_date
    description: The date on which any process or activity was ended
    todos:
    - add date string validation pattern
    comments:
    - We are using string representations of dates until all components of our ecosystem
      can handle ISO 8610 dates
    - The date should be formatted as YYYY-MM-DD
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: end_date
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: string
  qc_status:
    name: qc_status
    description: Stores information about the result of a process (ie the process
      of sequencing a library may have for qc_status of 'fail' if not enough data
      was generated)
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: qc_status
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: StatusEnum
  qc_comment:
    name: qc_comment
    description: Slot to store additional comments about laboratory or workflow output.
      For workflow output it may describe the particular workflow stage that failed.
      (ie Failed at call-stage due to a malformed fastq file).
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: qc_comment
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: string
  has_failure_categorization:
    name: has_failure_categorization
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: has_failure_categorization
    owner: MetatranscriptomeAssembly
    domain_of:
    - PlannedProcess
    range: FailureCategorization
    multivalued: true
    inlined: true
    inlined_as_list: true
  id:
    name: id
    description: A unique identifier for a thing. Must be either a CURIE shorthand
      for a URI or a complete URI
    notes:
    - 'abstracted pattern: prefix:typecode-authshoulder-blade(.version)?(_seqsuffix)?'
    - a minimum length of 3 characters is suggested for typecodes, but 1 or 2 characters
      will be accepted
    - typecodes must correspond 1:1 to a class in the NMDC schema. this will be checked
      via per-class id slot usage assertions
    - minting authority shoulders should probably be enumerated and checked in the
      pattern
    examples:
    - value: nmdc:mgmag-00-x012.1_7_c1
      description: https://github.com/microbiomedata/nmdc-schema/pull/499#discussion_r1018499248
    from_schema: https://w3id.org/nmdc/nmdc
    structured_aliases:
      workflow_execution_id:
        literal_form: workflow_execution_id
        predicate: NARROW_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
      data_object_id:
        literal_form: data_object_id
        predicate: NARROW_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
    rank: 1000
    identifier: true
    alias: id
    owner: MetatranscriptomeAssembly
    domain_of:
    - NamedThing
    range: uriorcurie
    required: true
    pattern: ^[a-zA-Z0-9][a-zA-Z0-9_\.]+:[a-zA-Z0-9_][a-zA-Z0-9_\-\/\.,]*$
    structured_pattern:
      syntax: '{id_nmdc_prefix}:wfmtas-{id_shoulder}-{id_blade}{id_version}$'
      interpolated: true
  name:
    name: name
    description: A human readable label for an entity
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: name
    owner: MetatranscriptomeAssembly
    domain_of:
    - PersonValue
    - NamedThing
    - Protocol
    range: string
  description:
    name: description
    description: a human-readable description of a thing
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    slot_uri: dcterms:description
    alias: description
    owner: MetatranscriptomeAssembly
    domain_of:
    - ImageValue
    - NamedThing
    - Protocol
    range: string
  alternative_identifiers:
    name: alternative_identifiers
    description: A list of alternative identifiers for the entity.
    from_schema: https://w3id.org/nmdc/nmdc
    rank: 1000
    alias: alternative_identifiers
    owner: MetatranscriptomeAssembly
    domain_of:
    - MetaboliteIdentification
    - NamedThing
    range: uriorcurie
    multivalued: true
    pattern: ^[a-zA-Z0-9][a-zA-Z0-9_\.]+:[a-zA-Z0-9_][a-zA-Z0-9_\-\/\.,\(\)\=\#]*$
  type:
    name: type
    description: the class_uri of the class that has been instantiated
    notes:
    - makes it easier to read example data files
    - required for polymorphic MongoDB collections
    examples:
    - value: nmdc:Biosample
    - value: nmdc:Study
    from_schema: https://w3id.org/nmdc/nmdc
    see_also:
    - https://github.com/microbiomedata/nmdc-schema/issues/1048
    - https://github.com/microbiomedata/nmdc-schema/issues/1233
    - https://github.com/microbiomedata/nmdc-schema/issues/248
    structured_aliases:
      workflow_execution_class:
        literal_form: workflow_execution_class
        predicate: NARROW_SYNONYM
        contexts:
        - https://bitbucket.org/berkeleylab/jgi-jat/macros/nmdc_metadata.yaml
    rank: 1000
    slot_uri: rdf:type
    designates_type: true
    alias: type
    owner: MetatranscriptomeAssembly
    domain_of:
    - EukEval
    - FunctionalAnnotationAggMember
    - PeptideQuantification
    - ProteinQuantification
    - MobilePhaseSegment
    - PortionOfSubstance
    - MagBin
    - MetaboliteIdentification
    - GenomeFeature
    - FunctionalAnnotation
    - AttributeValue
    - NamedThing
    - OntologyRelation
    - FailureCategorization
    - Protocol
    - CreditAssociation
    - Doi
    range: uriorcurie
    required: true
class_uri: nmdc:MetatranscriptomeAssembly

</code></pre>
</details>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link", "hide_page_title"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
<script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
<script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
<script src="../javascripts/tablesort.js"></script>
<script src="https://unpkg.com/mermaid@10.9.0/dist/mermaid.min.js"></script><script>mermaid.initialize({});</script></body>
</html>