
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://microbiomedata.github.io/nmdc-schema/identifiers/">
      
      
        <link rel="prev" href="../schema-validation/">
      
      
        <link rel="next" href="../typecode-to-class-map/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Identifiers - NMDC Schema Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#identifiers-in-nmdc" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="NMDC Schema Documentation" class="md-header__button md-logo" aria-label="NMDC Schema Documentation" data-md-component="logo">
      
  <img src="../images/NMDC_logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NMDC Schema Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Identifiers
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/microbiomedata/nmdc-schema" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="NMDC Schema Documentation" class="md-nav__button md-logo" aria-label="NMDC Schema Documentation" data-md-component="logo">
      
  <img src="../images/NMDC_logo.png" alt="logo">

    </a>
    NMDC Schema Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/microbiomedata/nmdc-schema" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Funding and Acknowledgements
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../collaboration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Metadata_Documentation_Overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Metadata Standards Documentation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../maintaining-the-schema/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Maintaining the NMDC Schema
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../schema_element_deprecation_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Schema element deprecation guide
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../schema-validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    NMDC Schema Validation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Identifiers
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Identifiers
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#curies-prefixed-ids" class="md-nav__link">
    <span class="md-ellipsis">
      CURIEs - prefixed IDs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    <span class="md-ellipsis">
      Examples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-insdc-biosamples-example-curie-biosamplesamea2397676" class="md-nav__link">
    <span class="md-ellipsis">
      1. INSDC BioSamples (Example CURIE: BIOSAMPLE:SAMEA2397676)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-gold-identifiers-example-curie-goldgp0119849" class="md-nav__link">
    <span class="md-ellipsis">
      2. GOLD identifiers (Example CURIE: GOLD:Gp0119849)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identifiers-for-ontology-terms-and-function-descriptors" class="md-nav__link">
    <span class="md-ellipsis">
      identifiers for ontology terms and function descriptors
    </span>
  </a>
  
    <nav class="md-nav" aria-label="identifiers for ontology terms and function descriptors">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kegg" class="md-nav__link">
    <span class="md-ellipsis">
      KEGG
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#recommended-ids-for-use-within-nmdc" class="md-nav__link">
    <span class="md-ellipsis">
      Recommended IDs for use within NMDC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ids-minted-for-use-within-nmdc" class="md-nav__link">
    <span class="md-ellipsis">
      IDs minted for use within NMDC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#annotation-identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation identifiers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reuse-vs-minting-new-ids" class="md-nav__link">
    <span class="md-ellipsis">
      Reuse vs minting new IDs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#identifier-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      Identifier mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-details-on-legacy-identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      Additional details on legacy identifiers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#mixs-term-identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      MIxS term identifiers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#identifiers-and-semantic-web-uris" class="md-nav__link">
    <span class="md-ellipsis">
      Identifiers and semantic web URIs
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../typecode-to-class-map/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Typecode to class map
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../migrations/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Version Migration Guides
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualizations/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Visualizations
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Validation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../developer-docs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Developer Docs
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#curies-prefixed-ids" class="md-nav__link">
    <span class="md-ellipsis">
      CURIEs - prefixed IDs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    <span class="md-ellipsis">
      Examples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-insdc-biosamples-example-curie-biosamplesamea2397676" class="md-nav__link">
    <span class="md-ellipsis">
      1. INSDC BioSamples (Example CURIE: BIOSAMPLE:SAMEA2397676)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-gold-identifiers-example-curie-goldgp0119849" class="md-nav__link">
    <span class="md-ellipsis">
      2. GOLD identifiers (Example CURIE: GOLD:Gp0119849)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#identifiers-for-ontology-terms-and-function-descriptors" class="md-nav__link">
    <span class="md-ellipsis">
      identifiers for ontology terms and function descriptors
    </span>
  </a>
  
    <nav class="md-nav" aria-label="identifiers for ontology terms and function descriptors">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kegg" class="md-nav__link">
    <span class="md-ellipsis">
      KEGG
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#recommended-ids-for-use-within-nmdc" class="md-nav__link">
    <span class="md-ellipsis">
      Recommended IDs for use within NMDC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ids-minted-for-use-within-nmdc" class="md-nav__link">
    <span class="md-ellipsis">
      IDs minted for use within NMDC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#annotation-identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation identifiers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reuse-vs-minting-new-ids" class="md-nav__link">
    <span class="md-ellipsis">
      Reuse vs minting new IDs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#identifier-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      Identifier mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-details-on-legacy-identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      Additional details on legacy identifiers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#mixs-term-identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      MIxS term identifiers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#identifiers-and-semantic-web-uris" class="md-nav__link">
    <span class="md-ellipsis">
      Identifiers and semantic web URIs
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="identifiers-in-nmdc">Identifiers in NMDC</h1>
<p>Identifiers are crucial for the NMDC, both for any data objects <em>created</em> (aka minted) and for any external objects
that are being <em>referenced</em> in NMDC.</p>
<p>Examples of entities that require identifiers:</p>
<ul>
<li>Samples</li>
<li>Data objects (e.g. sequence files)</li>
<li>Taxa (e.g. NCBITaxon or GTDB)</li>
<li>Genes, Proteins</li>
<li>Sequences (e.g. genome/transcriptome)</li>
<li>Ontology terms and other descriptors<ul>
<li>functional orthologs, e.g. KEGG.orthology (KO) terms</li>
<li>pathways, e.g. KEGG.pathway, MetaCyc, GO</li>
<li>reactions/activities: KEGG, MetaCyc</li>
<li>chemical entities: CHEBI, CHEMBL, INCHI, ...</li>
<li>sequence feature types: SO, Rfam</li>
</ul>
</li>
</ul>
<p>Identifiers should be:</p>
<ul>
<li>Permanent</li>
<li>Unique</li>
<li>Resolvable</li>
<li>Opaque</li>
</ul>
<p>See <a href="https://www.ncbi.nlm.nih.gov/pubmed/28662064">McMurry et al, PMID:28662064</a> for more desiderata.</p>
<h2 id="curies-prefixed-ids">CURIEs - prefixed IDs</h2>
<p>Following McMurry et al. we adopt the use of <em>prefixed identifiers</em></p>
<p>The syntax is:</p>
<pre><code>Prefix:LocalId
</code></pre>
<p>Examples:</p>
<ul>
<li>GO:0008152</li>
<li>BIOSAMPLE:SAMEA2397676</li>
<li>DOI:10.1038/nbt1156</li>
</ul>
<p>These prefixed identifiers are also known as CURIEs (Compact URIs). There is
a <a href="https://www.w3.org/TR/curie">W3C specification</a> for these.</p>
<p>All prefixes should be registered with at least one standard identifier prefix system.  If a prefix is not already
registered, please open a ticket at Bioregistry.io to request registration.  If an entity is registered in multiple 
registries (possibly with differing syntax or case), we recommend using the bioregistry.io registry as the primary
source of truth.</p>
<p>Popular choices for identifier registry services include:</p>
<ul>
<li><a href="https://bioregistry.io">https://bioregistry.io</a></li>
<li><a href="http://n2t.net">http://n2t.net</a></li>
<li><a href="http://identifiers.org">http://identifiers.org</a></li>
</ul>
<h2 id="examples">Examples</h2>
<h4 id="1-insdc-biosamples-example-curie-biosamplesamea2397676">1. INSDC BioSamples (Example CURIE: <code>BIOSAMPLE:SAMEA2397676</code>)</h4>
<table>
<thead>
<tr>
<th>Description</th>
<th>Registry Entry</th>
<th>Resolvable Link</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Bioregistry.io</strong></td>
<td><a href="https://bioregistry.io/registry/biosample">https://bioregistry.io/registry/biosample</a></td>
<td><a href="https://bioregistry.io/biosample:SAMEA2397676">https://bioregistry.io/biosample:SAMEA2397676</a></td>
</tr>
<tr>
<td><strong>Identifiers.org</strong></td>
<td><a href="https://registry.identifiers.org/registry/biosample">https://registry.identifiers.org/registry/biosample</a></td>
<td><a href="https://identifiers.org/BIOSAMPLE:SAMEA2397676">https://identifiers.org/BIOSAMPLE:SAMEA2397676</a></td>
</tr>
<tr>
<td><strong>N2T.net</strong></td>
<td></td>
<td><a href="http://n2t.net/BIOSAMPLE:SAMEA2397676">http://n2t.net/BIOSAMPLE:SAMEA2397676</a></td>
</tr>
</tbody>
</table>
<h4 id="2-gold-identifiers-example-curie-goldgp0119849">2. GOLD identifiers (Example CURIE: <code>GOLD:Gp0119849</code>)</h4>
<table>
<thead>
<tr>
<th>Description</th>
<th>Registry Entry</th>
<th>Resolvable Link</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Bioregistry.io</strong></td>
<td><a href="https://bioregistry.io/registry/gold">https://bioregistry.io/registry/gold</a></td>
<td><a href="https://bioregistry.io/gold:Gp0119849">https://bioregistry.io/gold:Gp0119849</a></td>
</tr>
<tr>
<td><strong>Identifiers.org</strong></td>
<td><a href="https://registry.identifiers.org/registry/gold">https://registry.identifiers.org/registry/gold</a></td>
<td><a href="https://identifiers.org/GOLD:Gp0119849">https://identifiers.org/GOLD:Gp0119849</a></td>
</tr>
<tr>
<td><strong>N2T.net</strong></td>
<td>N/A</td>
<td><a href="http://n2t.net/GOLD:Gp0119849">http://n2t.net/GOLD:Gp0119849</a></td>
</tr>
</tbody>
</table>
<h3 id="identifiers-for-ontology-terms-and-function-descriptors">identifiers for ontology terms and function descriptors</h3>
<p>Most of the ontologies we use are in OBO. All OBO IDs are prefixed using the ontology ID space. 
The list of ID spaces can be found on: <a href="http://obofoundry.org">http://obofoundry.org</a>, and use a special resolver service called "PURL" 
(Permanent URL) to resolve the ID to a URL.</p>
<p>For example the ID/CURIE <code>ENVO:00002007</code> represents the class <code>sediment</code> and is expanded to a URI
of <a href="http://purl.obolibrary.org/obo/ENVO_00002007">http://purl.obolibrary.org/obo/ENVO_00002007</a></p>
<h4 id="kegg">KEGG</h4>
<p>KEGG is actually a set of databases, each with its own prefix, usually of form <code>KEGG.$database</code>, e.g.</p>
<ul>
<li><a href="https://registry.identifiers.org/registry/kegg.orthology">KEGG.ORTHOLOGY</a> (aka KO), e.g. KEGG.ORTHOLOGY:K00001</li>
<li><a href="https://registry.identifiers.org/registry/kegg.compound">KEGG.COMPOUND</a>, e.g. KEGG.COMPOUND:C12345</li>
</ul>
<h2 id="recommended-ids-for-use-within-nmdc">Recommended IDs for use within NMDC</h2>
<p>The NMDC Schema is annotated with the set of IDs, ordered by preference, that are allowed to act as primary keys 
for instances of each class. For example the class <a href="https://microbiomedata.github.io/nmdc-metadata/docs/OrthologyGroup">OrthologyGroup</a> has a description of the IDs allowed 
on the class web page, the first listed is <a href="https://registry.identifiers.org/registry/kegg.orthology">KEGG.ORTHOLOGY</a>.  The full URL for each is in the jsonld context file, 
<a href="...">jsonschema/nmdc.context.jsonld</a>.</p>
<p>The underlying yaml looks like this:</p>
<pre><code class="language-yaml">  orthology group:
    is_a: functional annotation term
    description: &gt;-
      A set of genes or gene products in which all members are orthologous
    id_prefixes:
      - KEGG.ORTHOLOGY  ## KO number
      - EGGNOG
      - PFAM
      - TIGRFAM
      - SUPFAM
      - PANTHER.FAMILY
    exact_mappings:
      - biolink:GeneFamily
</code></pre>
<h2 id="ids-minted-for-use-within-nmdc">IDs minted for use within NMDC</h2>
<p>The NMDC Schema specifies legal identifiers for all of its classes. All data instances/records that are intended for 
upload into the NMDC metadata store must have an <code>id</code> field that follows this specification. Ids that are minted
at NMDC must match this abstract pattern:</p>
<pre><code>nmdc:&lt;type-code&gt;-&lt;shoulder&gt;-&lt;blade&gt;&lt;.version&gt;&lt;_locus&gt;
</code></pre>
<p>The abstract pattern has six parts, delimited by hyphens (unless otherwise specified):</p>
<ol>
<li>
<p><code>nmdc</code>: All NMDC identifiers will begin with this static prefix.</p>
</li>
<li>
<p><code>&lt;typecode&gt;</code>: An alphabetical code with a 1:1 correspondence to a class from the NMDC Schema. Answers the question "of what class is the data record that bears this <code>id</code>"? Must consist of 1 to 6 lower case letters, although a minimum of 3 letters is suggested. The <em>type code</em> portion of an NMDC <code>id</code> must match the regular expression <code>[a-z]{1,6}</code>.</p>
</li>
<li>
<p><code>&lt;shoulder&gt;</code>: A code that indicates what organization minted the identifier. Shoulder values must be zero to six lower case letters, flanked by one digit on either side. Answers the question "what organization minted this <code>id</code>"? The central identifier endpoint, hosted at LBL, uses the shoulder 00. Should organizations like JGI or EMSL need to mint identifiers in bulk, they would be assigned other shoulders, so that <code>id</code> values aren't reused. The <em>shoulder</em> portion of an NMDC <code>id</code> must match the regular expression <code>[0-9][a-z]{0,6}[0-9]</code>.</p>
</li>
<li>
<p><code>&lt;blade&gt;</code>: The fully unique part of the identifier under a given type code and shoulder namespace. The <em>shoulder</em> and <em>blade</em> together make up the <em>key</em> of the identifier. The blade is an alphanumeric string of open-ended length with at least one character, following the regular expression: <code>[A-Za-z0-9]+</code>.</p>
</li>
<li>
<p><code>&lt;.version&gt;</code>: Differentiates multiple iterations of a workflow. The delimiter used to separate the <em>version</em> from the <em>blade</em> and everything before it is a dot (<code>.</code>). The <em>version</em> is a potentially repeating alphanumeric pattern with a minimum length of 1 character. The <em>version</em> portion of an NMDC <code>id</code> must match the regular expression <code>(\.[A-Za-z0-9]+)*</code>.</p>
</li>
<li>
<p><code>&lt;_locus&gt;</code>: Indicates the contig on which a genomic feature is found, along with its start and end coordinates. Delimited from the rest of the <code>id</code> by an underscore (<code>_</code>). The <em>locus</em> part, if present, must have at least one character from the set off uppercase letters, lower case letters, digits, underscores (<code>_</code>), dots (<code>.</code>) and hyphens (<code>-</code>). The regular expression that the locus will follow is: <code>_[A-Za-z0-9_\.-]+</code>.</p>
</li>
</ol>
<p>The per-part regular expression described above can be composed into one complete regular expression. Named capture groups have been used to tie in the part names.</p>
<pre><code>^(?&lt;prefix&gt;nmdc):(?&lt;typecode&gt;[a-z]{1,6})-(?&lt;shoulder&gt;[0-9][a-z]{0,6}[0-9])-(?&lt;blade&gt;[A-Za-z0-9]+)(?&lt;version&gt;(\.[A-Za-z0-9]+)*)(?&lt;locus&gt;_[A-Za-z0-9_\.-]+)?$
</code></pre>
<p>NMDC offers a central identifier minting <a href="https://api.microbiomedata.org/docs#/minter/mint_ids_pids_mint_post">endpoint</a> in order to save data contributors the trouble of hand-crafting <code>id</code>s.
The possibility of decentralized (or offline) minting of <code>id</code>s by trusted organizations has also been anticipated. 
<code>id</code> component 3 below (the shoulder) is used to indicate the organization that minted an <code>id</code>. LBL, which hosts 
the <code>id</code> minting endpoint will use one shoulder value. If another organization, like JGI or EMSL, needs to bulk-create
<code>id</code>s outside of the central identifier minting endpoint, they would use different shoulders, to be determined by the 
NMDC Schema and metadata team.</p>
<h2 id="annotation-identifiers">Annotation identifiers</h2>
<p>Both metaG and metaT analyses produce GFF3 files. See <a href="https://github.com/microbiomedata/nmdc-metadata/issues/184">issue 184</a> for more on how the GFF is modeled.</p>
<p>The main entity identifier used in NMDC is the <a href="https://microbiomedata.github.io/nmdc-metadata/docs/GeneProduct">gene product</a> 
ID.  This identifier is used in functioanl annotations. This is typically a protein encoded by a CDS, e.g.</p>
<pre><code>Ga0185794_41    GeneMark.hmm-2 v1.05    CDS     48      1037    56.13   +       0       ID=Ga0185794_41_48_1037;translation_table=11;start_type=ATG;product=5-methylthioadenosine/S-adenosylhomocysteine deaminase;product_source=KO:K12960;cath_funfam=3.20.20.140;cog=COG0402;ko=KO:K12960;ec_number=EC:3.5.4.28,EC:3.5.4.31;pfam=PF01979;superfamily=51338,51556
</code></pre>
<p>Note: when processing GFF column 9 values, NMDC first ensures that each ID in this field is correctly prefixed 
according to the NMDC Schema <code>id_prefixes</code> directives. In the example above, <code>KO:K12960</code> is translated to
<code>KEGG.OTHOLOGY:K12960</code> to make sure the data is compliant with the registered prefix authorities and thus the 
NMDC Schema.  This helps to ensure that the data is interoperable and can be used in a variety of tools and
services beyond NMDC.</p>
<h2 id="reuse-vs-minting-new-ids">Reuse vs minting new IDs</h2>
<p>In 2023 NMDC transitioned from reusing identifiers from other organizations to using NMDC minted identifiers as the 
primary identifier. </p>
<h2 id="identifier-mapping">Identifier mapping</h2>
<table>
<thead>
<tr>
<th style="text-align: left;">Identifier</th>
<th style="text-align: left;">Example</th>
<th style="text-align: left;">NMDC Schema Class</th>
<th style="text-align: left;">NMDC Schema Slot</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">gold:Gs*</td>
<td style="text-align: left;">gold:Gs0114675</td>
<td style="text-align: left;">Study</td>
<td style="text-align: left;">gold_study_identifiers</td>
</tr>
<tr>
<td style="text-align: left;">gold:Gb*</td>
<td style="text-align: left;">gold:Gb0110739</td>
<td style="text-align: left;">Biosample</td>
<td style="text-align: left;">gold_biosample_identifiers</td>
</tr>
<tr>
<td style="text-align: left;">emsl:*</td>
<td style="text-align: left;">emsl:63ca2f94-6647-11eb-ae93-0242ac130002</td>
<td style="text-align: left;">Biosample</td>
<td style="text-align: left;">emsl_biosample_identifiers</td>
</tr>
<tr>
<td style="text-align: left;">igsn:*</td>
<td style="text-align: left;">igsn:IEWFS001H</td>
<td style="text-align: left;">Biosample</td>
<td style="text-align: left;">igsn_biosample_identifiers</td>
</tr>
<tr>
<td style="text-align: left;">gold:Gp*</td>
<td style="text-align: left;">gold:Gp0452734</td>
<td style="text-align: left;">OmicsProcessing</td>
<td style="text-align: left;">gold_sequencing_project_identifiers</td>
</tr>
<tr>
<td style="text-align: left;">emsl:*</td>
<td style="text-align: left;">emsl:598506</td>
<td style="text-align: left;">OmicsProcessing</td>
<td style="text-align: left;">alternative_identifiers</td>
</tr>
</tbody>
</table>
<p>Some legacy data object identifiers were based on file md5sums, either with or without a prefix (nmdc, jgi, emsl). 
In some cases the legacy value can be found by removing the prefix and searching DataObject records on slot 
md5_checksum.  If you are having trouble finding information based on legacy identifiers 
please contact <a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#117;&#112;&#112;&#111;&#114;&#116;&#64;&#109;&#105;&#99;&#114;&#111;&#98;&#105;&#111;&#109;&#101;&#100;&#97;&#116;&#97;&#46;&#111;&#114;&#103;">&#115;&#117;&#112;&#112;&#111;&#114;&#116;&#64;&#109;&#105;&#99;&#114;&#111;&#98;&#105;&#111;&#109;&#101;&#100;&#97;&#116;&#97;&#46;&#111;&#114;&#103;</a>.</p>
<h2 id="additional-details-on-legacy-identifiers">Additional details on legacy identifiers</h2>
<p>Legacy metagenomics objects look like this:</p>
<pre><code class="language-yaml">      id: &quot;gold:Gp0108335&quot;
      name: &quot;Thawing permafrost microbial communities from the Arctic, studying carbon transformations - Permafrost 712P3D&quot;
      has_input:
        - &quot;gold:Gb0108335&quot;
      part_of:
        - &quot;gold:Gs0112340&quot;
      has_output:
        - &quot;jgi:551a20d30d878525404e90d5&quot;
      omics_type: Metagenome
      type: &quot;nmdc:OmicsProcessing&quot;
      add_date: &quot;30-OCT-14 12.00.00.000000000 AM&quot;
      mod_date: &quot;22-MAY-20 06.13.12.927000000 PM&quot;
      ncbi_project_name: &quot;Thawing permafrost microbial communities from the Arctic, studying carbon transformations - Permafrost 712P3D&quot;
      processing_institution: &quot;Joint Genome Institute&quot;
      principal_investigator_name: &quot;Virginia Rich&quot;
</code></pre>
<p>the linked data object uses a jgi prefix and an md5 hash</p>
<pre><code class="language-yaml">      id: &quot;jgi:551a20d30d878525404e90d5&quot;
      name: &quot;8871.1.114459.GCCAAT.fastq.gz&quot;
      description: &quot;Raw sequencer read data&quot;
      file_size_bytes: 17586370657
      type: &quot;nmdc:DataObject&quot;
</code></pre>
<p>Legacy metaproteomics objects look like this:</p>
<pre><code class="language-yaml">      id: &quot;emsl:404590&quot;
      name: &quot;FECB_21_5093B_01_23Dec14_Tiger_14-11-12&quot;
      description: &quot;High res MS with low res CID MSn&quot;
      part_of:
        - &quot;gold:Gs0110132&quot;
      has_output:
        - &quot;emsl:output_404590&quot;
      omics_type: Proteomics
      type: &quot;nmdc:OmicsProcessing&quot;
      instrument_name: &quot;VOrbiETD03&quot;
      processing_institution: &quot;Environmental Molecular Sciences Lab&quot;
</code></pre>
<p>and the output data objects are formed from these:</p>
<pre><code class="language-yaml">      id: &quot;emsl:output_404590&quot;
      name: &quot;output: FECB_21_5093B_01_23Dec14_Tiger_14-11-12&quot;
      description: &quot;High res MS with low res CID MSn&quot;
      file_size_bytes: 503296678
      type: &quot;nmdc:DataObject&quot;
</code></pre>
<p>the data objects use hashes (md5) prefixed with <code>nmdc</code>:</p>
<pre><code class="language-yaml">      name: &quot;404590_resultant.tsv&quot;
      description: &quot;Aggregation of analysis tools{MSGFplus, MASIC} results&quot;
      file_size_bytes: 10948480
      type: &quot;nmdc:DataObject&quot;
      id: &quot;nmdc:e0c70280a7a23c7c5cc1e589f72e896e&quot;
</code></pre>
<h2 id="mixs-term-identifiers">MIxS term identifiers</h2>
<p>We are working with the GSC to provide permanent IDs for MIxS terms. Note these terms are schema-level rather than
data-level. For now we place these in the NMDC namespaces, e.g <code>nmdc:alt</code></p>
<h2 id="identifiers-and-semantic-web-uris">Identifiers and semantic web URIs</h2>
<p>Using LinkML's default tooling, we produce a JSON-LD context with the schema:</p>
<ul>
<li><a href="https://github.com/microbiomedata/nmdc-schema/blob/main/project/jsonschema/nmdc.schema.json">jsonschema/nmdc.context.jsonld</a></li>
</ul>
<p>When this is combined with schema-conformant JSON, RDF can be automatically created using the intended URIs</p>
<p>Please see: <a href="maintaining-the-schema">Maintaining identifiers</a>] for more information on 
developing and maintaining identifiers in the schema.</p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link", "hide_page_title"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
        <script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
      
        <script src="../javascripts/tablesort.js"></script>
      
    
  </body>
</html>